### DISCUSSION TITLE
How to get custom layout?

### DISCUSSION THREAD

**Discussion by pboling** [janko/rodauth-rails#295] [Q&A] [answered]
In my last question I was confused about the flash messages not rendering... the reason turns out to be that my custom layout isn't working.

I feel like I already did this before, but I can't figure out how if I did.  My rodauth-rails views are being rendered in my Rails' default "applciation" layout, which is highly unsuitable for my rodauth-rails use-case.  I see that the render plugin is configured within Rodauth-rails, but I'm not sure how to override that config.

I have a "parent" controller for my rodauth-rails authenticated routes on the Rails side, and I've set `layout "roda_api"` there, but it doesn't apply to the roda-handled routes of rodauth proper.

I'll try directly overriding the `plugin :render` next...

**Comment by pboling** [comment]
Can't get it to work.  I see:
```
  Rendering layout layouts/application.html.erb
  Rendering rodauth/login.html.erb within layouts/application
```

I tried:
```
  plugin :render, layout: "../../views/layouts/my_api"
```
And:
```
  plugin :render, layout: "my_api"
```
And a copy of the rodauth-rails internal line, swapping in the name of the alternate layout:
```
  plugin :render, :escape=>true, :check_paths=>true, :views=>'app/views', :layout=>'layouts/my_api.html'
```
My layout designations appear to have no effect at all.  No errors, and no indication that I've changed anything at all.  The layout is simply ignored.

**Comment by janko** [comment, marked as answer]
rodauth-rails loads the `render` Roda plugin only to be able to render Rodauth's built-in views, which are in Tilt and within a Roda instance. It only gets the page HTML, and forwards it to a Rails controller, which renders the layout, because that's a Rails view template rendered in Rails view context.

You change the layout in `RodauthController`, just like you do for normal Rails views, see https://github.com/janko/rodauth-rails/wiki/Different-layout.

**Reply by pboling** [reply]
It works!  Thanks :)

**Reply by pboling** [reply]
For anyone else ending up here you might have, like I did, two relevant controllers.  I had one parent controller for my rodauth-rails routes, and another parent controller for all my routes that were internally protected by rodauth-rails' authentication.  I had to set the layout in both of them as they had no relationship to each other!
